### PR TITLE
filter_state: add shared pointer accessor

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: Allow access to mutable filter state objects via shared pointer. This allows copying objects to different streams (e.g. from downstream to upstream filter state, or from downstream to another internal downstream filter state). Read-only objects are locked to the stream to avoid mutation via another stream.
Risk Level: low
Testing: unit
Docs Changes: none
Release Notes: none

### CODE REVIEW

**Comment by lambdai** [comment]
According to https://github.com/envoyproxy/envoy/issues/14320

we should not add throw exception at worker thread and eventually remove the existing one.

What about starting to use `StatusOr` here?

**Comment by kyessenov** [comment]
`StatusOr` sounds fine to me. The error here is not runtime error, it's a programming error, in case someone treats filter state as mutable accidentally.

File: test/common/stream_info/filter_state_impl_test.cc [hunk 1]
```diff
@@ -77,6 +77,25 @@ TEST_F(FilterStateImplTest, Simple) {
...
+  EXPECT_EQ(0u, destruction_count);
+
+  auto obj = filter_state().getDataSharedMutableGeneric("test_name");
+  EXPECT_EQ(5, dynamic_cast<TestStoredTypeTracking*>(obj.get())->access());
+  EXPECT_EQ(1u, access_count);
+  EXPECT_EQ(0u, destruction_count);
+
+  resetFilterState();
+  EXPECT_EQ(1u, access_count);
+  EXPECT_EQ(0u, destruction_count);
```
Comment (KBaichoo) [inline test/common/stream_info/filter_state_impl_test.cc:96]: when will the destruction count be incremented for shared pointer version?  why doesn't resetFilterState() increment it?

**Review by KBaichoo — COMMENTED** [review]
Regarding exception use here, this seems like a rehash of https://github.com/envoyproxy/envoy/pull/4820#discussion_r227164718. I think going from that example (since this is almost a copy-paste of the case there) that this exception usage is fine. cc @envoyproxy/senior-maintainers in case this has changed.

File: test/common/stream_info/filter_state_impl_test.cc (hunk 1 above)
Comment (kyessenov) [reply test/common/stream_info/filter_state_impl_test.cc:96]: When the shared pointer is deleted on block exit. This allows sharing objects between two streams without one stream deleting it.

**Comment by kyessenov** [comment]
Thanks @KBaichoo . I agree that we should stay consistent and then, separately, revisit exception use. IMHO, we should have used leaned on C++ type system to enforce const-ness instead of doing runtime checks.

File: test/common/stream_info/filter_state_impl_test.cc (hunk 1 above)
Comment (KBaichoo) [reply test/common/stream_info/filter_state_impl_test.cc:96]: makes sense, mind adding a reset to obj shared_ptr, then checking the `destruction_count` is incremented as expect? Thanks

Comment (kyessenov) [reply test/common/stream_info/filter_state_impl_test.cc]: Done.

**Review by KBaichoo — DISMISSED** [review]
tsan failure is fixed in https://github.com/envoyproxy/envoy/pull/20262, but maybe the retest above will get lucky.

File: source/common/stream_info/filter_state_impl.cc [hunk 2]
```diff
@@ -79,6 +79,26 @@ FilterState::Object* FilterStateImpl::getDataMutableGeneric(absl::string_view da
   return current->data_.get();
 }
 
+std::shared_ptr<FilterState::Object>
+FilterStateImpl::getDataSharedMutableGeneric(absl::string_view data_name) {
```
Comment (mattklein123) [inline source/common/stream_info/filter_state_impl.cc:83]: This is the same function as above, right? Can't you just implement the one above in terms of this one?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
I agree that if the code were new I would use StatusOr instead of exceptions, but I don't think it's worth boiling the ocean here when the existing code already works like this.

Small comment about duplicated code though. Thank you.

/wait

File: source/common/stream_info/filter_state_impl.cc (hunk 2 above)
Comment (kyessenov) [reply source/common/stream_info/filter_state_impl.cc:83]: No, it has a different recursive call. Not sure how to templatize the recursion so copying seems simpler.

Comment (mattklein123) [reply source/common/stream_info/filter_state_impl.cc:83]: I'm probably missing something, but can't you just call the shared_ptr version and then just convert it to raw pointer after the recursion is complete?

Comment (kyessenov) [reply source/common/stream_info/filter_state_impl.cc:68]: Done.

**Pull request merged**